### PR TITLE
feat!: Rename inductive/co to least/greatest

### PR DESCRIPTION
**Pull request by RustanLeino**
This PR makes a syntax change in the language, to make _extreme_ (that is, _least_ or _greatest_) _predicates_ more uniform and more understandable.

| old syntax | new syntax |
|------------|------------|
| `inductive predicate` | `least predicate` |
| `copredicate` | `greatest predicate` |
| `inductive lemma` | `least lemma` |
| `colemma` | `greatest lemma` |

For now, the old syntax is still supported, but gives rise to deprecation warnings (with the default `/deprecation:1` command-line option).

The new syntax introduces _reserved keyword phrases_. In particular, the 4 pairs of words in the right-hand column of the table above are reserved keyword phrases, but `least` and `greatest` by themselves are not reserved keywords. That is, only when `least` or `greatest` is followed by `predicate` or `lemma` is it reserved. This allows a program to still use `least` and `greatest` as ordinary identifier names. A syntax-highlighting IDE can still be simplistic--it would highlight `least` and `greatest` only for occurrences that are followed by `predicate` or `lemma`.

## Three kinds of predicates

Dafny already had 3 kinds of predicates, but now the description of these align with the keyword (phrase)s used to declare them. This section summarizes that description. For more details, see this [tutorial](https://easychair.org/publications/download/vHsB) and its references.

Consider a definition

    ... predicate P(...) {
      ... P(...) ...
    }

If the first ellipsis is empty, then `P` is defined to be an ordinary predicate, that is, a boolean function that is verified to terminate. Given that it terminates, it evaluates to a unique result for any input (that satisfies the predicate's precondition).

If the first ellipsis is `least` or `greatest`, then `P` is defined as the least or greatest, respectively, solution to the following recurrence equation:

    P(...)    =    ... P(...) ...

For such a solution to exist, every occurrence of `P` on the right-hand side (that is, in the body of the Dafny definition) must appear in a positive position.

For every least or greatest predicate `P`, Dafny automatically adds a definition of a _prefix predicate_ `P#`. A prefix predicate has a special syntax, both in the `#` that ends its name and also in its _index_, which is singled out in square brackets between the name `P#` and the ordinary parameters. The value of `P#[n](...)` essentially denotes `n` applications of the right-hand side of the recurrence equation (starting from `false` for a least predicate and from `true` for a greatest predicate). (I say "essentially", because `n` is by default of type `ORDINAL`.)

The common way to prove properties of an ordinary predicate is to use a recursive `lemma`. This gives a proof by induction (over the termination metric of the lemma, which is typically the same as that of the predicate of interest). To prove properties of extreme predicates, Dafny provides corresponding extreme lemmas (`least lemma` and `greatest lemma`), which do their proofs by induction over the index of the predicate's prefix predicate. Dafny's`least lemma` and `greatest lemma` declarations provide the illusion of writing these proofs directly in terms of the extreme predicates.

## How the old syntax came to be

The phrase “inductive predicate” comes from type theory, and it is what Coq, Agda, Isabelle, F*, and others have adopted as their terminology. Up until now, Dafny had done so, too.

Dafny actually had co-inductive predicates before it had inductive predicates, and for these used the syntax `copredicate`.

One way to make `inductive predicate` and `copredicate` look more like the duals of each other that they are would have been to have renamed the latter `coinductive predicate`. However, I always find it confusing to explain "inductive predicates" to users. After all, isn’t every ordinary recursively defined predicate an “inductive” predicate?

Yes, an ordinary predicate is "inductive" in the sense that you can prove things about it using induction. But what sets an "inductive predicate" apart is that it is defined as _the least solution_ (of the recurrence equation above). Considering the corresponding prefix predicate, one can see how the definition of the predicate is "built up" from `false`, that is, its value is justified inductively through a proof tree. One can (and one does) apply induction to reason about this proof tree. But the definition of the predicate is a least solution. Therefore, it seems appropriate to use the name `least predicate`.

Analogously, using `greatest predicate` for what used to be `copredicate` offers both a better name and a more conspicuous duality with `least predicate`.

### CODE REVIEW

File: Source/Dafny/Dafny.atg [hunk 1]
```diff
@@ -1102,7 +1116,7 @@ DatatypeDecl<DeclModifierData dmod, ModuleDefinition/*!*/ module, out DatatypeDe
      List<DatatypeCtor/*!*/> ctors = new List<DatatypeCtor/*!*/>();
      IToken bodyStart = Token.NoToken;  // dummy assignment
      bool co = false;
-     CheckDeclModifiers(dmod, "Datatypes or codatatypes", AllowedDeclModifiers.None);
+     CheckDeclModifiers(dmod, "datatypes or codatatype", AllowedDeclModifiers.None);
```
Comment (davidcok) [inline Source/Dafny/Dafny.atg:1119]: should datatypes be datatype ?

**Review by davidcok — CHANGES_REQUESTED** [review]
When we were discussion the grammar design for least/greatest, I had the impression that one would just need to check for optional "least" or "greatest" before a "predicate" or "lemma" keyword. But it appears that Coco automatically makes them keywords, so you have to add them back into the set of identifiers. Is that right? Does every literal string used in the grammar automatically become reserved? [If so, I would not have argued for my position, though I prefer it, because it makes the grammar more complicated and maintenance more error-prone.]

File: Source/Dafny/Dafny.atg (hunk 1 above)
Comment (RustanLeino) [reply Source/Dafny/Dafny.atg:1119]: yep, done

**Comment by RustanLeino** [comment]
> it appears that Coco automatically makes them keywords, so you have to add them back into the set of identifiers. Is that right? Does every literal string used in the grammar automatically become reserved?

Yes, that's correct. Luckily, other uses of the scanner's `ident` went through the user-defined `Ident` (or, at least they do now), so this could be accomplished without much effort. I agree we'll have to keep an eye on this in the future; that is, that we should use `Ident` in `Dafny.atg` and not `ident` directly.

**Review by davidcok — APPROVED** [review]

**Pull request merged**